### PR TITLE
[mess up] remote_src of copy.py  supports recursive copy of directory

### LINKED ISSUE #14131: Allow copy module to work with recursive and remote_src
In Ansible 2.0, when using:

``` yml
copy: src={{src_dir}} dest={{dest_dir}} remote_src=yes
```

Copy module fails with:
`"msg": "attempted to take checksum of directory: <remote source directory>"`

This happens with and without `/` slashes at the ends of both or either directory name, and also whether or not the destination directory exists or not.  Source directory always present and filled with files and subdirectories.

### PR DESCRIPTION
**Pull request by ptux**
##### SUMMARY

fixes #14131 

The followings will be supported.

* push remote src to remote dest with idempotence.
* supports check mode currently.
* supports symlink.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

module copy.py

##### ANSIBLE VERSION

```
ansible 2.7.0.dev0 (issue-14131 974076c7cb) last updated 2018/06/07 13:19:30 (GMT +900)
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['~/ansible/library']
  ansible python module location = ~/ansible/lib/ansible
  executable location = ~/ansible/bin/ansible
  python version = 3.6.5 (default, Apr 25 2018, 14:23:58) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.1)]

```

### CODE REVIEW

**Comment by jborean93** [comment]
Not sure if there is much use in adding recursive copying for remote_src if there is not idempotence or check mode. The user may as well just run a cp command. @abadger I believe you were planning on adding support for this at some point, do you have any feelings about this?

**Comment by ptux** [comment]
@jborean93 
thanks for your comment.
I was confused about idempotence.
I did not implement idempotence, because I thought this is copy, not sync.

I would add [check mode] later.

**Comment by jborean93** [comment]
@ptux no worries, `remote_src: no` is idempotent, it checks each file before copying to see if it needs to happen. This is part of the reason why `copy` is slower than just plain SSH commands and not everyone realises this.

**Comment by ptux** [comment]
> it checks each file before copying to see if it needs to happen.

this could happen by using filecmp module of python lib. 
But, it's slow and there is an alternate [synchronize].
 It is annoying.

@abadger
advise, please.

**Comment by abadger** [comment]
Copy need to be idempotent and should handle check_mode.  As jborean says,
it is the major reason to use the copy module instead of command: cp

**Comment by abadger** [comment]
Also note, filecmp uses file metadata (mode, ownership, etc) as part of
it's comparison.  That may not be appropriate for copy since the user can
set some of those as copy parameters.

**Comment by ptux** [comment]
I see.
I'll try to implement idempotence and check mode both.

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/68986/36/tests

**Comment by ptux** [comment]
@abadger 

Cloud you take a look at the latest commit?
please advise me if there are any problems.

**Comment by ptux** [comment]
Is there anybody who will review this PR?

File: lib/ansible/modules/files/copy.py [hunk 1]
```diff
@@ -78,7 +78,7 @@
     description:
       - If C(no), it will search for I(src) at originating/master machine.
       - If C(yes) it will go to the remote/target machine for the I(src). Default is C(no).
-      - Currently I(remote_src) does not support recursive copying.
+      - I(remote_src) support recursive copying.
```
Comment (abadger) [inline lib/ansible/modules/files/copy.py]: Add version here like:
```
- I(remote_src) supports recursive copying in version 2.7 and higher
```

**Comment by abadger** [comment]
* Needs integration tests.
* Does this handle permissions?  For instance:
```
- copy:
    remote_src: True
    src: /etc
    dest: ~/backup
    mode: u+rwX,go-rwx
```

**Comment by ptux** [comment]
thanks for your comments.

I'll add code to handle permissions.

File: lib/ansible/modules/files/copy.py (hunk 1 above)
Comment (ptux) [reply lib/ansible/modules/files/copy.py]: got it.

**Comment by ptux** [comment]
@abadger 

I was coding for handling `mode`,`owner`,`group`.
I noticed it's NOT a good idea that supports `mode` for **remote recursive copy**.
Which I mean who wants to grant SAME permission to dir and file.

So, I  suggest that make **remote recursive copy** support `owner` and `group` only.
Users cloud use `file` module to change permission for dir and file **separately**

Let me know your opinion, please.

**Comment by abadger** [comment]
Symbolic modes work.  And those are what we recommend for recursive

**Comment by ptux** [comment]
@abadger 

* push remote src to remote dest with idempotence.
* supports check mode currently.
* supports symlink.

review one more time, please.
tell me if there is something wrong or missing.

**Comment by ptux** [comment]
ready_for_review

**Comment by ptux** [comment]
@abadger 

I guess you have a lot going on.
just a friendly reminder.

**Comment by abadger** [comment]
@ptux thanks.  This still neds integration tests.  Probably somewhere around here: https://github.com/ansible/ansible/blob/devel/test/integration/targets/copy/tasks/tests.yml#L1160

**Comment by ptux** [comment]
@abadger 
I didn't know how and where to write the test code.
I'll try my best to do it.

**Comment by ptux** [comment]
@abadger 

I'll continue writing the test code.
cloud you take a look at it at this monment?

advise me if I'm going in the wrong  **DIRECTION**.

**Comment by mattclay** [comment]
CI failure in integration tests: https://app.shippable.com/github/ansible/ansible/runs/77323/46/tests

**Comment by ptux** [comment]
@abadger 
@IPvSean 
@mattclay 

ready to review!
anybody, please.

**Comment by ptux** [comment]
I messed up this PR.
I have created a clean one #43998.

**Pull request closed without merging**